### PR TITLE
Spring REST Docs

### PR DESCRIPTION
**Pull request by Antrakos**
Abstract classes and preferances. Geo zone example

### CODE REVIEW

File: web/src/main/java/edu/softserve/zoo/Error.java [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
 import edu.softserve.zoo.exceptions.ApplicationException;
+import org.apache.commons.lang3.StringUtils;
 
 public class Error {
 
     private String message;
 
     public Error(ApplicationException exception) {
-        this.message = exception.getReason().getMessage();
+        this.message = StringUtils.isEmpty(exception.getMessage())? exception.getReason().getMessage() : exception.getMessage();
```
Comment (ITsvetkoFF) [inline web/src/main/java/edu/softserve/zoo/Error.java]: leave as it was

File: web/src/test/java/edu/softserve/zoo/rest/docs/AbstractCrud.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+ */
+public abstract class AbstractCrud extends AbstractTest {
+    @Autowired
+    ObjectMapper objectMapper;
+
+    /**
+     * {@code GET} request test
+     */
+    protected void getOp() throws Exception {
+        RestDocumentationResultHandler document = documentPrettyPrintReqResp("get" + getName());
```
Comment (ITsvetkoFF) [inline web/src/test/java/edu/softserve/zoo/rest/docs/AbstractCrud.java]: get all request mappings

File: web/src/main/java/edu/softserve/zoo/controller/rest/WarehouseController.java [hunk 3]
```diff
@@ -32,22 +36,28 @@ protected WarehouseController() {
         return warehouseService;
     }
 
+    public static WarehouseDto getTestDto() {
```
Comment (ghost) [inline web/src/main/java/edu/softserve/zoo/controller/rest/WarehouseController.java]: there should not be test dto in controllers

File: web/src/test/java/edu/softserve/zoo/rest/docs/DocsGenerationTest.java [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+    private RestDocumentationResultHandler documentPrettyPrintReqResp(String useCase) {
+        return document(useCase,
+                preprocessRequest(prettyPrint()),
+                preprocessResponse(prettyPrint()));
+    }
+
+    @Test
+    public void run() throws Exception {
+        StringBuilder documentation = new StringBuilder();
+        documentation.append("= Zoo REST API Documentation\n")
```
Comment (ghost) [inline web/src/test/java/edu/softserve/zoo/rest/docs/DocsGenerationTest.java]: move to properties file or somewhere else. you don't need to append this by stringbuilder

File: web/src/test/java/edu/softserve/zoo/rest/docs/DocsGenerationTest.java [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+                .append("| `200 OK`\n")
+                .append("| Standard response for successful HTTP requests. The actual response will depend on the request method used. In a GET request, the response will contain an entity corresponding to the requested resource. In a POST request, the response will contain an entity describing or containing the result of the action.\n")
+                .append("\n")
+                .append("| `404 Not Found`\n")
+                .append("| The requested resource could not be found but may be available again in the future. Subsequent requests by the client are permissible.\n")
+                .append("|===\n")
+                .append("\n")
+                .append("[[resources]]\n")
+                .append("= Resources\n");
+        for (Map.Entry<RequestMappingInfo, HandlerMethod> mapping :
```
Comment (ghost) [inline web/src/test/java/edu/softserve/zoo/rest/docs/DocsGenerationTest.java:114]: move to separate methods.

**Comment by Antrakos** [comment]
image
That's the problem.

**Pull request merged**